File: cpp/199_BinaryTreeRightSideView.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        vector<int> ret;
        if(root == NULL) return ret;
        queue<TreeNode*> qTree;
        qTree.push(root);
        qTree.push(NULL);
        while(qTree.size()){
            auto t = qTree.front();
            qTree.pop();
            if(t != NULL){
                auto n = qTree.front();
                if(n == NULL){
                    ret.push_back(t->val);
                }
                if(t->left) qTree.push(t->left);
                if(t->right) qTree.push(t->right);
            }else{
                if(qTree.size()) qTree.push(NULL);
            }
        }
        return ret;
    }
};

File: js/199_BinaryTreeRightSideView.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var rightSideView = function(root) {
    var ret = [];
    if(root === null) return ret;
    var queue = [];
    queue.push(root);
    queue.push(null);
    while(queue.length){
        var t = queue.shift();
        if( t !== null){
            var n = queue[0];
            if( n === null) ret.push(t.val);
            if(t.left) queue.push(t.left);
            if(t.right) queue.push(t.right);
        }else{
            if(queue.length) queue.push(null);
        }
    }
    return ret;
};